api/system/v1alpha1/zone_insight_helpers.go

```diff
@@ -64,13 +72,14 @@ func (x *ZoneInsight) UpdateSubscription(s generic.Subscription) error {
 	if !ok {
 		return errors.Errorf("invalid type %T for ZoneInsight", s)
 	}
-	i, old := x.GetSubscription(kdsSubscription.Id)
-	if old != nil {
-		x.Subscriptions[i] = kdsSubscription
-	} else {
-		x.finalizeSubscriptions()
-		x.Subscriptions = append(x.Subscriptions, kdsSubscription)
+	for i, sub := range x.GetSubscriptions() {
+		if sub.GetId() == kdsSubscription.Id {
+			x.Subscriptions[i] = kdsSubscription
+			return nil
+		}
 	}
+	x.finalizeSubscriptions()
```

Comment:
I thought that we discussed that we want to drop this call

Reply:
Yeah. Wasn't sure whether to put it in this PR since it causes somewhat different problems but I've added it

Reply:
@lobkovilya

Reply:
But `finalizeSubscription` will close old subscriptions quicker than finalizer, right? So if 1 CP instance stopped ungracefully and DPP reconnects to another CP, then you'll immediately get old subscription closed

Reply:
But I don't think they necessarily stay disconnected. In [my original PR](https://github.com/kumahq/kuma/pull/8144) I describe how I believe in the ungraceful case, `ZoneInsightSink` can in fact _unset_ the disconnected time because it doesn't check the current state.

Reply:
Right, why did we drop this idea? We should probably check `DisconnectTime` before updating the subscription to avoid the unset.

Reply:
I'm not sure, I've come around to "let finalizer handle all disconnects". It's somehow odd to have two different sources of truth for disconnection state, one in UpdateSubscription and one in finalizer. @jakubdyszkiewicz